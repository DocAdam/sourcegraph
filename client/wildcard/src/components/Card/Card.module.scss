:root {
    --card-bg: var(--color-bg-1);
    --card-border-color: var(--gray-04);
}

.card {
    position: relative;
    display: flex;
    flex-direction: column;
    min-width: 0; // See https://github.com/twbs/bootstrap/pull/22740#issuecomment-305868106
    word-wrap: break-word;
    background-color: var(--card-bg);
    background-clip: border-box;
    border-width: 1px;
    border-style: solid;
    border-color: var(--card-border-color);
    border-radius: var(--border-radius);

    > hr {
        margin-right: 0;
        margin-left: 0;
    }
}

.card-interactive {
    cursor: pointer;
    &:hover:not(:global(.focus)):not(:focus):not(:active) {
        // TODO: Improve implementation,to avoid visual card size transitions
        border-width: 2px;
        border-color: var(--blue);
    }

    &:focus,
    &:active {
        box-shadow: 0 0 0 2px var(--primary-2);
        border-color: var(--white);
    }
}
