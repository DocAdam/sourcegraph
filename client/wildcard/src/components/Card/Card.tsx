import classNames from 'classnames'
import React, { ReactNode } from 'react'

import styles from './Card.module.scss'
import { CardBody, CardHeader, CardSubtitle, CardTitle } from './components'
import { CARD_VARIANTS } from './constants'

interface CardProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'title'> {
    header?: ReactNode
    title?: ReactNode
    subtitle?: ReactNode
    /**
     * Interactive variant, shows blue border on hover and focus
     */
    variant?: typeof CARD_VARIANTS[number]
}

/**
 * Card Element
 */
export const Card: React.FunctionComponent<CardProps> = ({
    header,
    title,
    subtitle,
    children,
    className,
    variant = 'default',
    ...attributes
}) => {
    const cardHeader = header && <CardHeader>{header}</CardHeader>
    const cardTitle = title && <CardTitle>{title}</CardTitle>
    const cardSubtitle = subtitle && <CardSubtitle>{subtitle}</CardSubtitle>
    const cardBody =
        title || subtitle ? (
            <CardBody>
                {cardTitle}
                {cardSubtitle}
                {children}
            </CardBody>
        ) : (
            children
        )

    return (
        <div
            className={classNames(styles.card, className, variant === 'interactive' && styles.cardInteractive)}
            {...attributes}
        >
            {cardHeader}
            {cardBody}
        </div>
    )
}
