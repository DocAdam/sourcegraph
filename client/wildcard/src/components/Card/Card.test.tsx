import { render, screen } from '@testing-library/react'
import React from 'react'

import { Card } from './Card'

describe('Card', () => {
    it('renders card correctly', () => {
        const { asFragment } = render(
            <Card title="card title" subtitle="card subtitle" variant="interactive">
                <div>Card Body</div>
            </Card>
        )

        expect(screen.getByText(/card title/)).toHaveClass('cardTitle')
        expect(screen.getByText(/card subtitle/)).toHaveClass('cardSubtitle')
        expect(asFragment()).toMatchSnapshot()
    })
})
