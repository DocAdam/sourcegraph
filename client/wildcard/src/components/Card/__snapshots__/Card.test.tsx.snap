// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders card correctly 1`] = `
<DocumentFragment>
  <div
    class="card cardInteractive"
  >
    <div
      class="cardBody"
    >
      <div
        class="cardTitle"
      >
        card title
      </div>
      <div
        class="cardSubtitle"
      >
        card subtitle
      </div>
      <div>
        Card Body
      </div>
    </div>
  </div>
</DocumentFragment>
`;
